server/src/main/java/org/opensearch/index/store/CompositeDirectory.java

```diff
@@ -0,0 +1,354 @@
+        return files;
+    }
+
+    /**
+     * Removes an existing file in the directory.
+     * Currently deleting only from local directory as files from remote should not be deleted as that is taken care by garbage collection logic of remote directory
+     * @param name the name of an existing file.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void deleteFile(String name) throws IOException {
+        ensureOpen();
+        ensureFileNotDeleted(name);
+        logger.trace("Composite Directory[{}]: deleteFile() called {}", this::toString, () -> name);
+        if (FileTypeUtils.isTempFile(name)) {
+            localDirectory.deleteFile(name);
+        } else if (Arrays.asList(listAll()).contains(name) == false) {
+            throw new NoSuchFileException("File " + name + " not found in directory");
+        } else {
+            pendingDeletes.add(name);
+        }
+    }
+
+    /**
+     * Returns the byte length of a file in the directory.
+     * Throws {@link NoSuchFileException} or {@link FileNotFoundException} in case file is not present locally and in remote as well
+     * @param name the name of an existing file.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public long fileLength(String name) throws IOException {
+        ensureOpen();
+        ensureFileNotDeleted(name);
+        logger.trace("Composite Directory[{}]: fileLength() called {}", this::toString, () -> name);
+        long fileLength;
+        Path key = localDirectory.getDirectory().resolve(name);
+        if (FileTypeUtils.isTempFile(name) || fileCache.get(key) != null) {
+            try {
+                fileLength = localDirectory.fileLength(name);
+                logger.trace(
+                    "Composite Directory[{}]: fileLength of {} fetched from Local - {}",
+                    this::toString,
+                    () -> name,
+                    () -> fileLength
+                );
+            } finally {
+                fileCache.decRef(key);
+            }
+        } else {
+            fileLength = remoteDirectory.fileLength(name);
+            logger.trace(
+                "Composite Directory[{}]: fileLength of {} fetched from Remote - {}",
+                this::toString,
+                () -> name,
+                () -> fileLength
+            );
+        }
+        return fileLength;
+    }
+
+    /**
+     * Creates a new, empty file in the directory and returns an {@link IndexOutput} instance for
+     * appending data to this file.
+     * @param name the name of the file to create.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public IndexOutput createOutput(String name, IOContext context) throws IOException {
+        ensureOpen();
+        // If file was deleted earlier, removing it from the deleted list
+        if (pendingDeletes.contains(name)) {
+            pendingDeletes.remove(name);
+        }
```

Comment:
Could not find any file for which this behaviour happens currently, but this check in implemented currently in [FSDirectory](https://github.com/apache/lucene/blob/750a7c4d3b3e174023404bf363861dae31413901/lucene/core/src/java/org/apache/lucene/store/FSDirectory.java#L211) as well. 

So added it here as well in case we encounter this for tests that we haven't tried out yet.